test/e2e/volumes.go

```diff
@@ -0,0 +1,360 @@
+// configured ports from it. The returned pod should be used to get the server
+// IP address and create appropriate VolumeSource.
+// The function may return nil - the test should be skipped in this case.
+func startVolumeServer(client *client.Client, config VolumeTestConfig) *api.Pod {
+	podClient := client.Pods(config.namespace)
+
+	portCount := len(config.serverPorts)
+	serverPodPorts := make([]api.ContainerPort, portCount)
+
+	for i := 0; i < portCount; i++ {
+		portName := fmt.Sprintf("%s-%d", config.prefix, i)
+
+		serverPodPorts[i] = api.ContainerPort{
+			Name:          portName,
+			ContainerPort: config.serverPorts[i],
+			Protocol:      api.ProtocolTCP,
+		}
+	}
+
+	By(fmt.Sprint("creating ", config.prefix, " server pod"))
+	serverPod := &api.Pod{
+		TypeMeta: api.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1beta3",
+		},
+		ObjectMeta: api.ObjectMeta{
+			Name: config.prefix + "-server",
+			Labels: map[string]string{
+				"role": config.prefix + "-server",
+			},
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:       config.prefix + "-server",
+					Image:      config.serverImage,
+					Privileged: true,
+					Ports:      serverPodPorts,
+				},
+			},
+		},
+	}
+	if _, err := podClient.Create(serverPod); err != nil {
+		// Skip the test when 'privileged' containers are not allowed
+		if strings.Contains(err.Error(), "spec.containers[0].privileged: forbidden 'true'") {
+			By(fmt.Sprint("Skipping ", config.prefix, " test, which is supported only on on cluster with --allow_privileged=True"))
+			return nil
+		}
+		// Report real error otherwise
+		Failf("Failed to create %s pod: %v", serverPod.Name, err)
+	}
+
+	expectNoError(waitForPodRunningInNamespace(client, serverPod.Name, config.namespace))
+
+	By("locating the server pod")
+	pod, err := podClient.Get(serverPod.Name)
+	expectNoError(err, "Cannot locate the server pod %v: %v", serverPod.Name, err)
+
+	By("sleeping a bit to give the server time to start")
+	time.Sleep(20 * time.Second)
+	return pod
+}
+
+// Clean both server and client pods.
+func volumeTestCleanup(client *client.Client, config VolumeTestConfig) {
+	By(fmt.Sprint("cleaning the environment after ", config.prefix))
+
+	defer GinkgoRecover()
+
+	podClient := client.Pods(config.namespace)
+
+	// ignore all errors, the pods may not be even created
+	podClient.Delete(config.prefix + "-client")
+	podClient.Delete(config.prefix + "-server")
+
+	if config.namespace != "default" {
+		client.Namespaces().Delete(config.namespace)
+	}
+}
+
+// Start a client pod using given VolumeSource (exported by startVolumeServer())
+// and check that the pod sees the data from the server pod.
+func testVolumeClient(client *client.Client, config VolumeTestConfig, volume api.VolumeSource, expectedContent string) {
+	By(fmt.Sprint("starting ", config.prefix, " client"))
+	podClient := client.Pods(config.namespace)
+
+	clientPod := &api.Pod{
```

Comment:
nit: The use of both podClient and clientPod here is marginally confusing.  Perhaps just 'pod' for the latter?